### ISSUE TITLE
Import Caffe models to deeplearn.js

### ISSUE THREAD

**Issue by chaosmail** [PAIR-code/deeplearnjs#267]
To make it as simple as possible for the user, it would be great to parse both `deploy.prototxt` (model layout) and `*.caffemodel` (model weights) files directly within the browser. As both files are simply protos this should be possible. After playing around, I found that the actual model architecture is embedded in the original `*.caffemodel` file itself - hence we need to only load this file. Heck, even the training mean is included!

The Caffe proto definition can be found here: https://github.com/BVLC/caffe/blob/master/src/caffe/proto/caffe.proto

The following steps would be necessary:

- [x] Parse the `caffe.proto` to Typescript module using dcodeIO (https://github.com/dcodeIO/ProtoBuf.js/#pbts-for-typescript)

```sh
$ pbjs -t static-module -w commonjs -o compiled.js file1.proto file2.proto
$ pbts -o compiled.d.ts compiled.js
```
An alternative would be to serve the `caffe.proto` file and parse it at runtime (discussion of trade-offs: https://github.com/dcodeIO/ProtoBuf.js/#reflection-vs-static-code), which would not require a build step. I would prefer a build/compile step during dev than more dependencies at runtime. What do you guys think?

- [x] Add `protobuf.js` as a dependency. Maybe even the light version is enough for us.

- [x] Fetch the binary `*.caffemodel` file, parse it to NetParameter and convert it to a deeplearn.js model

regarding 1.:

As a proof-of-concept, I converted the `caffe.proto` file into a `caffe.json` proto (using `cli/bin/pbjs -t json tmp/caffe.proto > tmp/caffe.json`) and successfully loaded the `squeezenet.caffemodel` (both weights and structure) from the original file only. Here is a working demo:

```html
<html>
  <head>
    <!-- https://github.com/dcodeIO/protobuf.js -->
    <script src="libs/protobuf.min.js"></script>
  </head>
  <body>
    <script>
  
    function fetchArrayBuffer(uri) {
      var req = new Request(uri);
      return fetch(req).then(function(res) {
        return res.arrayBuffer();
      });
    }

    protobuf.load("caffe.json", function(err, root) {
      
      // Read caffe proto
      var caffe = root.lookup('caffe');
 
      // Fetch caffemodel
      fetchArrayBuffer('squeezenet_v1.1.caffemodel').then(function(res){

        var model = caffe.NetParameter.decode(new Uint8Array(res));
        console.log(model);
      });
    });
    </script>
  </body>
</html>
```

And here the parsed model plus weights:

![bildschirmfoto vom 2017-10-26 21-53-56](https://user-images.githubusercontent.com/2969388/32074316-3f3f3744-ba99-11e7-8190-55d481cc53f5.png)

Looking promising so far for fully featured Caffe model support!

**Comment by chaosmail** [comment]
just for reference: if we would have to parse binaryproto files that store e.g. the full training mean - here is a working snippet: 

```js
// Load binaryproto
fetchArrayBuffer('models/age_net/mean.binaryproto').then(function(res){
  var mean = caffe.BlobProto.decode(new Uint8Array(res));
  console.log(mean);
});
```

**Comment by chaosmail** [comment]
considering `deploy.prototxt`: if for any reason we want to load the model structure only - here is a working (but super hacky) snippet:

```html
<html>
  <head>
    <script src="libs/protobuf.min.js"></script>
  </head>
  <body>
    <script>

    function fetchText(uri) {
      var req = new Request(uri);
      return fetch(req).then(function(res) {
        return res.text();
      });
    }

    // dirty hack
    function prototxtToJson(raw, level) {
      level = level || 0;

      var json = {};
      var match;

      if (level == 0) {
        var regexVal = /(?:^|\n)(\w+):\s"*([\w/.]+)"*/gi;
        var regexObj = /(?:^|\n)(\w+)\s\{([\S\s]*?)\n\}/gi;
      }
      else {
        let indent = '(?:^|\\n)\\s{' + level + '}';
        let key = '(\\w+)';
        var regexVal = new RegExp(indent + key + '\\s*:\\s*"*([\\w/.]+)"*', "gi");
        var regexObj = new RegExp(indent + key + '\\s*\\{\\s*\\n([\\s\\S]*?)\\n\\s{' + level + '}\\}', "gi");
      }

      while (match = regexVal.exec(raw)) {
        let key = match[1];
        let value = match[2];
        if (json[key] !== undefined) {
          if (Array.isArray(json[key])) {
            json[key].push(value);
          }
          else {
            json[key] = [json[key]];
            json[key].push(value);
          }
        }
        else {
          json[match[1]] = value;
        }
      }

      while (match = regexObj.exec(raw)) {
        let key = match[1];
        let value = prototxtToJson(match[2], level + 2);

        if (json[key] !== undefined) {
          if (Array.isArray(json[key])) {
            json[key].push(value);
          }
          else {
            json[key] = [json[key]];
            json[key].push(value);
          }
        }
        else {
          json[key] = value;
        }
      }

      return json;
    }
  
    protobuf.load("caffe.json", function(err, root) {
      
      // Read caffe proto
      var caffe = root.lookup('caffe');
      
      // Fetch model definition
      fetchText('models/squeezenet/deploy.prototxt').then(function(res){

        var modelDef = caffe.NetParameter.create(prototxtToJson(res));
        console.log(modelDef);
      });
    });

    </script>
  </body>
</html>
```

**Comment by chaosmail** [comment]
I am making good progress so far (https://github.com/chaosmail/deeplearnjs/tree/caffe).

I am now at a point, where I need to transform the CaffeModel layers into a (static) Graph of operations for the deeplearn.js model. Otherwise I would have to parse the CaffeModel structure on every forward pass.

My idea was to build a DAG when I parse the model, and then push the input Array through the Graph in the infer function (instead of building the DAG in the infer function as in `Squeezenet::infer`).

You guys have any thoughts about this?

**Comment by chaosmail** [comment]
What is the intention of returning the pre-softmax logits in the `Squeezenet::infer` function? I think it would be more generic if we return the last layer of the model and store all activations for access of logits, etc.

**Comment by chaosmail** [comment]
It seems like that the information about the phase (TRAIN, TEST) is missing (or incorrect) in most caffemodel files and hence we need the `deploy.prototxt` files as well.

**Comment by nsthorat** [comment]
Sorry about the delay in responding! Glad you are making progress.

One thought about protobuf.js: it seems pretty well supported. The official google jspb is a huge pain in the ass to work with (you have to use bazel, and I want to avoid that). The one thing I'm concerned about is bloating our library with a dependency, something to keep in mind.

When you say Graph, do you mean our Graph class? I think we should think about an intermediate representation that we can share between Caffe and other models we may want to serialize. I actually think we need to do this at the `NDArrayMath` layer, a DAG sounds good for now. I think in the future we're going to want to share this with how we record a tape of operations in the new eager style of differentiation.

The intention of returning pre-softmax is that often you don't actually need the softmax normalized probabilities as a common use case is just use argmax to compute the highest class score, or use it in a transfer learning setting. Softmax itself is super expensive on the GPU, so we avoid it and let the user do the normalization if they want probabilities.

Hope this helps.

**Comment by chaosmail** [comment]
### protobuf.js

I strongly agree. However, I was pleasantly surprised of how much has happend since I last used protobuf.js (like one year ago). The compiled `caffe.proto` file (`caffe.js`) is still huge, but protobuf dependency seems to be minimal and interop with typescript works amazingly well.

### Graph

Exactly, I meant the Graph class of deeplearn.js. I strongly agree that this should be done at the NDArrayMath, as general as possible - however until now I didn't find a good way how this could be implemented (reusing the Graph class) without some major changes. However, I didn't invest much time until now, as I focused on model/weights loading and parsing first.

### Squeezenet::infer / Squeezenet::predict

I see, very nice! However, I think a more general way to implement this (once the DAG operations graph is available), is to specfiy the output layer in the inference method, such as `model::predict(input: Array3D, untilLayer: string = 'prob')`.

If softmax is expensive and not needed for TopK, then maybe it could be removed from Squeezenet::getTopKClasses in line https://github.com/PAIR-code/deeplearnjs/blob/3ff21da115e00e757a0bb600cb4940b6fc9edd25/models/squeezenet/squeezenet.ts#L152.

**Comment by chaosmail** [comment]
Here is my current status of the Prototxt Parser using Parsimmon (here is a working example: https://github.com/chaosmail/deeplearnjs/blob/prototxt/src/contrib/caffe_parser.ts). However, this requires an update of Parsimmon Types (here is my PR https://github.com/DefinitelyTyped/DefinitelyTyped/pull/21626).

**Comment by chaosmail** [comment]
#514 #517
